### PR TITLE
Model state channels last checkpoint patch

### PR DESCRIPTION
**Pull request by pzharrington**
# PhysicsNeMo Pull Request

## Description
Follow-up to #1597. Addresses the following issue with channels-last models loaded in FSDP-only mode:

- Channels-last conv weights were silently permuted on non-rank-0 ranks during FSDP checkpoint restart. `set_model_state_dict(broadcast_from_rank0=True)` invokes `dist.broadcast` per tensor, which accepts a channels_last-strided sender (its contiguity check passes for that format) but transfers bytes in storage order; receivers allocate `torch.empty(shape, dtype, device)` in standard NCHW layout, so the CL bytes land in NCHW positions and 4-D conv weights end up permuted on every rank except rank 0. The fix is a small load-side helper that makes positive-dim tensors standard-contiguous on rank 0 before handing the state dict to DCP, so the broadcast becomes layout-consistent across ranks; this matches the structure of the existing optim-state remap, which sidesteps the same hazard.

Also improves the cross-compatibility between checkpoints saved in single-process mode vs. single-/multi-process FSDP and single-/multi-process FSDP+ShardTensor mode by having the optimizer/model state loading infer channels_last format from the model in question rather than the tensors saved in the checkpoint itself.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

**Review by CharlelieLrt — APPROVED** [review]
No major comment, looks good to me!

File: physicsnemo/utils/checkpoint.py [hunk 1]
```diff
@@ -237,6 +265,66 @@ def _fsdp_uses_flat_param_optim(model: torch.nn.Module | None) -> bool:
...
+            flat_param._shapes,
+            flat_param._strides,
+            flat_param._contiguities,
+        ):
+            if contig:
+                continue
+            # CL / CL3D both have channel stride == 1 (channel is the
+            # innermost / fastest-varying dim in NHWC / NDHWC storage).
+            if len(shape) in (4, 5) and len(stride) == len(shape) and stride[1] == 1:
+                cl_fqns.add((prefix + fqn).removeprefix("_orig_mod."))
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply physicsnemo/utils/checkpoint.py]: Clarified the intent with a helper

File: examples/weather/stormcast/test_training.py [hunk 2]
```diff
@@ -305,6 +312,38 @@ def test_checkpoint_integrity(
...
+    t2 = trainer.Trainer(cfg_diffusion.copy())
+    net2 = t2.net
+    opt2 = t2.optimizer
+
+    options = StateDictOptions(full_state_dict=True)
+    (params1, opt_params1) = get_state_dict(net1, opt1, options=options)
+    (params2, opt_params2) = get_state_dict(net2, opt2, options=options)
+
+    assert set(params1.keys()) == set(params2.keys())
+    assert set(opt_params1.keys()) == set(opt_params2.keys())
```
Comment (jleinonen) [inline examples/weather/stormcast/test_training.py]: Maybe copy here the assert messages from line 296 and 312.
